python/ray/data/_internal/execution/operators/map_data_processor.py

```diff
@@ -0,0 +1,317 @@
+import functools
+import itertools
+from enum import Enum
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+from ray.data._internal.block_batching.block_batching import batch_blocks
+from ray.data._internal.execution.interfaces.task_context import TaskContext
+from ray.data._internal.output_buffer import BlockOutputBuffer
+from ray.data.block import Block, BlockAccessor, DataBatch
+from ray.data.context import DataContext
+
+Row = Dict[str, Any]
+MapTransformFnData = Union[Block, Row, DataBatch]
+MapTransformCallable = Callable[
+    [Iterable[MapTransformFnData], TaskContext], Iterable[MapTransformFnData]
+]
+BlockTransformCallable = Callable[[Iterable[Block], TaskContext], Iterable[Block]]
+RowTransformCallable = Callable[[Iterable[Row], TaskContext], Iterable[Row]]
+BatchTransformCallable = Callable[
+    [Iterable[DataBatch], TaskContext], Iterable[DataBatch]
+]
```

Comment:
Let's use parametrized types for better readability.

At the top we can declare just a single `TransformCallable`
```python
T = TypeVar("T")
TransformCallable = Callable[[Iterable[T], TaskContext], Iterable[T]]
```

And then in the method signatures we can specify `TransformCallable[Row]`, `TransformCallable[Block]` and `TransformCallable[DataBatch]` when appropriate.

https://stackoverflow.com/questions/46382170/how-can-i-create-my-own-parameterized-type-in-python-like-optionalt

Reply:
nice suggestion.